### PR TITLE
[Merged by Bors] - feat: add API for finite adele rings

### PR DESCRIPTION
**Pull request by kbuzzard**
Let R be a Dedekind domain with field of fractions K. This PR does two things:

1) makes the finite adeles of K/R into a K-algebra

2) turns `finiteAdeleRing` into a type rather than a subring (because it's most definitely an important object of study in its own right, e.g. it needs a topology, and I don't want to end up in the same kind of hell that we had with rings of integers before they were a type).

Note that this is a leaf file which I intend to develop a fair bit more in the near future.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

- [x] depends on: #13020
- [x] depends on: #13111

### CODE REVIEW

File: Mathlib/Algebra/Order/Field/Canonical/Defs.lean [hunk 1]
```diff
@@ -34,3 +34,32 @@ instance (priority := 100) CanonicallyLinearOrderedSemifield.toCanonicallyLinear
...
+
+open Function
+
+-- this makes `mul_lt_mul_left`, `mul_pos` etc work on `ℤₘ₀`
+instance {α : Type*} [Mul α] [Preorder α] [CovariantClass α α (· * ·) (· < ·)]:
+    PosMulStrictMono (WithZero α) where
+  elim := @fun
+    | ⟨(x : α), hx⟩, 0, (b : α), _ => by
+        rw [mul_zero]
+        exact WithZero.zero_lt_coe _
```
Comment (faenuccio) [inline Mathlib/Algebra/Order/Field/Canonical/Defs.lean:46]:
```suggestion
        simpa only [mul_zero] using WithZero.zero_lt_coe _
```

File: Mathlib/RingTheory/DedekindDomain/FiniteAdeleRing.lean [hunk 2]
```diff
@@ -134,10 +135,16 @@ section AlgebraInstances
 instance : Algebra K (K_hat R K) :=
   (by infer_instance : Algebra K <| ∀ v : HeightOneSpectrum R, v.adicCompletion K)
```
Comment (faenuccio) [inline Mathlib/RingTheory/DedekindDomain/FiniteAdeleRing.lean:137]: Why not marking it simp?

File: Mathlib/RingTheory/DedekindDomain/FiniteAdeleRing.lean [hunk 3]
```diff
@@ -266,6 +273,35 @@ theorem one : (1 : K_hat R K).IsFiniteAdele := by
   convert finite_empty
 #align dedekind_domain.prod_adic_completions.is_finite_adele.one DedekindDomain.ProdAdicCompletions.IsFiniteAdele.one
 
+open scoped DiscreteValuation
+
```
Comment (faenuccio) [inline Mathlib/RingTheory/DedekindDomain/FiniteAdeleRing.lean:277]: Can you add a bit of doc, either here or at the top of the file (or at both!) with these new additions?

File: Mathlib/RingTheory/DedekindDomain/FiniteAdeleRing.lean [hunk 4]
```diff
@@ -274,20 +310,18 @@ open ProdAdicCompletions.IsFiniteAdele
...
+    mul_mem' := mul
+    one_mem' := one
+    add_mem' := add
+    zero_mem' := zero
+    algebraMap_mem' := algebraMap'
+  } : Subalgebra K (K_hat R K))
+#align dedekind_domain.finite_adele_ring DedekindDomain.FiniteAdeleRing
+
+instance : CommRing (FiniteAdeleRing R K) := Subalgebra.toCommRing _
+
```
Comment (faenuccio) [inline Mathlib/RingTheory/DedekindDomain/FiniteAdeleRing.lean:324]: Why do you want to get rid of the lemma `mem_finiteAdeleRing_iff`?

File: Mathlib/Algebra/Order/Field/Canonical/Defs.lean (hunk 1 above)
Comment (kbuzzard) [reply Mathlib/Algebra/Order/Field/Canonical/Defs.lean]: (this is a comment on one of the dependent PRs and was dealt with there)

File: Mathlib/RingTheory/DedekindDomain/FiniteAdeleRing.lean (hunk 3 above)
Comment (kbuzzard) [reply Mathlib/RingTheory/DedekindDomain/FiniteAdeleRing.lean:278]: I've added "We give this (finite adele) ring a K-algebra structure". Is this enough?

File: Mathlib/RingTheory/DedekindDomain/FiniteAdeleRing.lean (hunk 4 above)
Comment (kbuzzard) [reply Mathlib/RingTheory/DedekindDomain/FiniteAdeleRing.lean:325]: I've changed `FiniteAdeleRing` from a term to a type, so the lemma no longer makes any sense as it stands, and I couldn't think of what to replace it with.

Comment (kbuzzard) [reply Mathlib/RingTheory/DedekindDomain/FiniteAdeleRing.lean:329]: Coe lemma added!

Comment (faenuccio) [reply Mathlib/RingTheory/DedekindDomain/FiniteAdeleRing.lean:329]: Thanks, definitely LGTM! IMHO this PR is ready to go.

**Comment by riccardobrasca** [comment]
!bench

**Comment by riccardobrasca** [comment]
Thanks!

bors merge

**Pull request closed without merging**